cumulus_library/statistics/psm.py

```diff
@@ -0,0 +1,247 @@
+        )
+        self.queries.append(dataset_query)
+
+    def generate_psm_analysis(self, cursor: object, schema: str):
+        """Runs PSM statistics on generated tables"""
+        df = cursor.execute(f"select * from {self.config.target_table}").as_pandas()
+        symptoms_dict = self._get_symptoms_dict(self.config.classification_json)
+        for dependent_variable, codes in symptoms_dict.items():
+            df[dependent_variable] = df["code"].apply(lambda x: 1 if x in codes else 0)
+        df = df.drop(columns="code")
+        # instance_count present but unused for PSM if table contains a count_ref input
+        # (it's intended for manual review)
+        df = df.drop(columns="instance_count", errors="ignore")
+
+        columns = []
+        if self.config.join_cols_by_table is not None:
+            for table_key in self.config.join_cols_by_table:
+                for column in self.config.join_cols_by_table[table_key][
+                    "included_cols"
+                ]:
+                    if len(column) == 2:
+                        columns.append(column[1])
+                    else:
+                        columns.append(column[0])
+
+        for column in columns:
+            encoded_df = pandas.get_dummies(df[column])
+            df = pandas.concat([df, encoded_df], axis=1)
+            df = df.drop(column, axis=1)
+        df = df.reset_index()
+
+        try:
+            psm = PsmPy(
+                df,
+                treatment=self.config.dependent_variable,
+                indx=self.config.primary_ref,
+                exclude=[],
+            )
+            # This function populates the psm.predicted_data element, which is required
+            # for things like the knn_matched() function call
+            psm.logistic_ps(balance=True)
+            print(psm.predicted_data)
+            # This function populates the psm.df_matched element
+            # TODO: flip replacement to false after increasing sample data size
+            psm.knn_matched(
+                matcher="propensity_logit",
+                replacement=True,
+                caliper=None,
+                drop_unmatched=True,
+            )
+            print(psm.df_matched)
+        except ZeroDivisionError:
+            print(
+                "Encountered a divide by zero error during statistical graph generation. Try increasing your sample size."
+            )
+        except ValueError:
+            print(
+                "Encountered a value error during KNN matching. Try increasing your sample size."
+            )
+
+    def prepare_queries(self, cursor: object, schema: str):
+        self._create_covariate_table(cursor, schema)
+
+    def execute_queries(
+        self,
+        cursor: object,
+        schema: str,
+        verbose: bool,
+        drop_table: bool = False,
+    ):
+        super().execute_queries(cursor, schema, verbose, drop_table)
+        self.comment_queries()
+        self.write_queries()
```

Comment:
Isn't this a debugging step that writes the queries to disk? A :scissors: case?

Should `write_queries` just call `comment_queries` itself?

Reply:
yeah this is just a debugging step, will trim.

i could go either way on making this embedded or not - i think it depends on the complexity of the queries. feel free to talk me into just doing it always.